File: 20-Valid-Parentheses/20-Valid-Parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character,Character> map = new HashMap<>();
        Stack<Character> stack = new Stack<>();
        map.put(')','(');
        map.put('}','{');
        map.put(']','[');
        for(int i = 0; i < s.length(); i++){
            if(map.containsKey(s.charAt(i))){
                if(!stack.isEmpty() && map.get(s.charAt(i)) == stack.peek()){
                    stack.pop();
                }
                else{
                    return false;
                }
            } else {
                stack.push(s.charAt(i));
            }
        }
        return stack.isEmpty();
    }
}

File: 20-Valid-Parentheses/20-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        m = { ')' : '(', ']' : '[', '}' : '{'}
        stack = []
        for i in s:
            if i in m:
                if stack and m[i] == stack[-1]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(i)
        return len(stack) == 0
        #brute force
        # opening = ['[', '{', '(']
        # stack = []
        # for i in s:
        #     if len(stack) == 0:
        #         stack.append(i)
        #     else:
        #         if i in opening:
        #             stack.append(i)
        #         else:
        #             if stack[-1] == '(' and i == ')':
        #                 stack.pop()
        #             else:
        #                 if stack[-1] == '[' and i == ']':
        #                     stack.pop()
        #                 else:
        #                     if stack[-1] == '{' and i == '}':
        #                         stack.pop()
        #                     else:
        #                         stack.append(i)
        # return len(stack) == 0